File: server/filters/ccmap.c
#include "filters.h"

FILTER_SETUP( ccmap )
{
	char* copy = strdup(args_name);
	unsigned char in_cc;
	char *pt;
	if( callback->args.ccmap.out_cc[127] == 0 )
		memset(callback->args.ccmap.out_cc, 0xff, 128);
	pt = strtok(copy, ","); callback->args.ccmap.channel = strtoul( pt, NULL, 10 ) - 1;
	pt = strtok(NULL, ","); in_cc = strtoul( pt, NULL, 16 );
	pt = strtok(NULL, ","); callback->args.ccmap.out_cc[in_cc] = strtoul( pt, NULL, 16 );
	printf("ccmap %d CC %d to %d",
		callback->args.ccmap.channel,
		in_cc,
		callback->args.ccmap.out_cc[in_cc]
	);
	free(copy);
}
FILTER_CALLBACK( ccmap )
{
	unsigned char cur_state;
	unsigned char exp_state = 0xb0 + args->ccmap.channel;
	unsigned char cc = 0xff, val = 0xff;

	int a = 0;
	for( int b = 0; b < n_bytes; ++b )
	{
		unsigned char cur_state = scan_status(data, buf[b]);
		if( cur_state == exp_state && buf[b] < 0x80 )
		{
			if( cc == 0xff )
			{
				unsigned char out_cc = args->ccmap.out_cc[ buf[b] ];
				cc = out_cc < 0x80 ? out_cc : 0xfe;
			}
			else if( val == 0xff )
			{
				if( cc < 0x80 )
				{
					val = buf[b];
					out_buf[a++] = exp_state;
					out_buf[a++] = cc;
					out_buf[a++] = val;
				}
				val = 0xff;
				cc = 0xff;
			}
		}
		else
		{
			cc = 0xff;
			val = 0xff;
		}
	}
	return write_buffer( data->output_device->midi, out_buf, a );
}

File: rust/src/filters/ccmap.rs

pub struct CCMap { channel : u8, out_cc : [u8 ; 0xff] }
impl CCMap {
	pub fn new( args : String ) -> CCMap {
		let out_cc = [ 0u8 ; 0xff ];
		let channel = args.split(",").next().unwrap();

		let mut ccmap = CCMap {
			channel : parse_hex(channel),
			out_cc : out_cc,
		};
		ccmap.add_args( args );
		ccmap
	}
}
impl CallbackFn for CCMap {
	fn add_args( &mut self, args : String ) {
		let mut arg = args.split(",");
		self.channel = parse_hex(arg.next().unwrap()) - 1;
		let in_cc : usize = parse_hex(arg.next().unwrap()).try_into().unwrap();
		self.out_cc[in_cc] = parse_hex( arg.next().unwrap() );
		println!("ccmap {} CC {} to {}",
			self.channel,
			in_cc,
			self.out_cc[in_cc]
		);
	}
	fn callback(&self, data : &mut CallbackData, buf : &[u8]) -> usize {
		let exp_state : u8 = 0xb0 + self.channel;
		let mut cc : u8 = 0xff;
		let mut val : u8 = 0xff;

		let mut a = 0;
		for c in buf {
			let cur_state : u8 = data.output_device.scan_status(*c);
			if cur_state == exp_state && *c < 0x80 {
				if cc == 0xff {
					let s : usize = (*c).try_into().unwrap();
					let out_cc : u8 = self.out_cc[s];
					if out_cc < 0x80 {
						cc = out_cc;
					} else {
						cc = 0xfe;
					}
				} else if val == 0xff {
					if cc < 0x80 {
						val = *c;
						data.buffer[a] = exp_state; a += 1;
						data.buffer[a] = cc; a += 1;
						data.buffer[a] = val; a += 1;
					}
					val = 0xff;
					cc = 0xff;
				}
			} else {
				cc = 0xff;
				val = 0xff;
			}
		}
		data.output_device.send_buffer(&data.buffer[0..a]).unwrap()
	}
}
